### DISCUSSION TITLE
Failed to apply binding to property [DependencyPropertyDetails(Source)] on [Microsoft.UI.Xaml.Controls.Image]

### DISCUSSION THREAD

**Discussion by HakanL** [unoplatform/uno.toolkit.ui#1258] [Q&A]
How can I troubleshoot and fix this error I get after upgrading to Uno SDK 5.3 (from multi-head 5.1)? This is the error:
`Failed to apply binding to property [DependencyPropertyDetails(Source)] on [Microsoft.UI.Xaml.Controls.Image] (ImageSourceConverter cannot convert from DMXCore.DMXCore100.Models.MenuItem.)`
Call stack:
```
System.NotSupportedException: ImageSourceConverter cannot convert from DMXCore.DMXCore100.Models.MenuItem.
   at System.ComponentModel.TypeConverter.GetConvertFromException(Object value)
   at System.ComponentModel.TypeConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
   at Microsoft.UI.Xaml.Media.ImageSourceConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value) in /__w/1/s/src/Uno.UI/UI/Xaml/Media/ImageSourceConverter.cs:line 62
   at Uno.UI.DataBinding.BindingPropertyHelper.ConvertUsingTypeDescriptor(Type type, Object value) in /__w/1/s/src/Uno.UI/DataBinding/BindingPropertyHelper.cs:line 1345
   at Uno.UI.DataBinding.BindingPropertyHelper.Convert(Type propertyType, Object value) in /__w/1/s/src/Uno.UI/DataBinding/BindingPropertyHelper.cs:line 1389
   at Microsoft.UI.Xaml.Data.BindingExpression.SetTargetValueSafe(Object v, Boolean useTargetNullValue) in /__w/1/s/src/Uno.UI/DataBinding/BindingExpression.cs:line 700
```

Note that the `DMXCore.DMXCore100.Models.MenuItem` is my class object, it's just a class that has properties (like title, icon/avatar, etc) for a menu displayed in a WrapPanel with Uno Cards. I don't bind the class to any property, but I'm binding the Avatar Icon on the Card to a string property in that class that has the path of the icon to display. I didn't get these errors in earlier versions, but also, everything is working, the icons are displayed, and if I remove the binding then the error goes away (and also the icon). Unfortunately my project is fairly complex so it's not trivial to create a sample to reproduce this, I'm initially hoping someone can point me in the right direction of how I can troubleshoot the source of this error, and come up with a fix.

**Comment by MartinZikmund** [comment]
@HakanL can you please send a snippet of that relevant part of XAML that includes the binding? Also - it might be interesting to attach to `DataContextChanged` to see which layer gets set to that specific instance

**Reply by HakanL** [reply]
Sure, I can do that. Here's the ListView:
```
<ListView Grid.Row="1" Name="ListContainer" ItemsSource="{Binding MenuItems}"
	  Padding="8"
	  ItemTemplate="{StaticResource ContactGridViewTemplate}"
	  SelectionMode="None"
	  HorizontalAlignment="Center">
	<ListView.ItemContainerStyle>
		<Style TargetType="ListViewItem">
			<Setter Property="FocusVisualPrimaryBrush" Value="{StaticResource MyFocusVisualPrimaryBrush}"/>
			<Setter Property="FocusVisualSecondaryBrush" Value="{StaticResource MyFocusVisualSecondaryBrush}"/>
			<Setter Property="FocusVisualPrimaryThickness" Value="3" />
			<Setter Property="FocusVisualSecondaryThickness" Value="1" />
		</Style>
	</ListView.ItemContainerStyle>

	<ItemsControl.ItemsPanel>
		<ItemsPanelTemplate>
			<controls:WrapPanel VerticalSpacing="2" HorizontalSpacing="2" />
		</ItemsPanelTemplate>
	</ItemsControl.ItemsPanel>
</ListView>
```
And the template is defined like this:
```
<DataTemplate x:Key="ContactGridViewTemplate" x:DataType="models:MenuItem">
	<utu:Card HeaderContent="{x:Bind Name, Mode=OneWay}" Width="344" Margin="1,3,0,3"
		SubHeaderContent="{x:Bind Description, Mode=OneWay}"
		AvatarContent="{x:Bind FullIconPath}"
		Style="{StaticResource AvatarOutlinedCardStyle}" IsTabStop="False" BorderBrush="Black" BorderThickness="2"
		Tapped="Card_Tapped" Holding="Card_Holding" RightTapped="Card_RightTapped" AvatarContentTemplate="{StaticResource AvatarContentTemplateFixed40x40}">
	</utu:Card>
</DataTemplate>
```
`FullIconPath` is a string that is pointing to an embedded resource like this `ms-appx:///Assets/Icons/Icon.png`.
I captured the `DataContextChanged` on the ListView and it gets called once, with my view model, which is correct (and it has a list property called `MenuItems` of type `DMXCore.DMXCore100.Models.MenuItem`. Right after that event I get the NotSupportedExceptions, once per "card".

**Reply by MartinZikmund** [reply]
@HakanL very interesting, I don't see anything problematic in the snippet... Would you be willing to create an issue with a zip repro? It seems it indeed is a bug on our side.

**Reply by HakanL** [reply]
I'll try to break it out into a standalone app for repro.

**Reply by HakanL** [reply]
Simple repro repo: https://github.com/HakanL/UnoApp10
I get the errors both on my RPi 4 device, and in WSL2 on my Windows dev machine. Appreciate any insight!

**Reply by jeromelaban** [reply]
@kazo0 the issue seems to come from `MaterialAvatarOutlinedCardStyle`:

https://github.com/unoplatform/uno.toolkit.ui/blob/57206f4508f259b902d11bb5b9965c82803223bc/src/library/Uno.Toolkit.Material/Styles/Controls/v1/Card.xaml#L509

And this is:

https://github.com/unoplatform/uno.toolkit.ui/blob/57206f4508f259b902d11bb5b9965c82803223bc/src/library/Uno.Toolkit.Material/Styles/Controls/v1/Card.xaml#L61-L65

Where the binding does not seem to be changing its path in any way.

**Reply by kazo0** [reply]
@jeromelaban Seems that we need to avoid allowing the DataContext to be inherited for the XContent and XContentTemplate properties in the Card.

Similar to how it's done by ContentControl: https://github.com/unoplatform/uno/blob/0cf4db7ec540adb2942f42943790f638bad6cde3/src/Uno.UI/UI/Xaml/Controls/ContentControl/ContentControl.cs#L100-L106

Not sure if we would be able to do the same thing in Toolkit? Since `FrameworkPropertyMetadata` is an Uno-only thing right?

But I'm a little confused because we are forwarding these properties to the inner `ContentPresenter`: https://github.com/unoplatform/uno.toolkit.ui/blob/57206f4508f259b902d11bb5b9965c82803223bc/src/library/Uno.Toolkit.Material/Styles/Controls/v2/Card.xaml#L649-L659

So I would have though that the ContentPresenter would properly set its ContentTemplate's DataContext to Content

**Reply by MartinZikmund** [reply]
Yes, I would expect the `DataContext` in the template to correctly be the `Content` - when both `Content` and `ContentTemplate` are specified, the `ContentPresenter` should use the template to display the `Content`, which is then used as the `DataContext` of the template. I would expect the same works correctly in WinUI, right? Did this work in previous versions of Uno? If yes, then it might be a regression

**Reply by Xiaoy312** [reply]
@kazo0 you can do so:
```cs
		public static DependencyProperty TestProperty { get; } = DependencyProperty.Register(
			nameof(Test),
			typeof(string),
			typeof(UnreleatedTestClassForOtherThing),
#if !HAS_UNO
			new PropertyMetadata(default(string))
#else
			new FrameworkPropertyMetadata(default(string), FrameworkPropertyMetadataOptions.ValueDoesNotInheritDataContext | FrameworkPropertyMetadataOptions.AffectsMeasure)
#endif
		);
```

**Reply by HakanL** [reply]
Any suggestions on how I can work around/fix this issue? I'm still getting this on Skia with Uno.SDK 5.7.0-dev.137, although the type of exception has changed:
```
19:45:27.778 ERR [BindingExpression] Failed to apply binding to property [DependencyPropertyDetails(Source)] on [Microsoft.UI.Xaml.Controls.Image] (ImageSourceConverter cannot convert from DMXCore.DMXCore100.ViewModels.HomeViewModel.)
System.NotSupportedException: ImageSourceConverter cannot convert from DMXCore.DMXCore100.ViewModels.HomeViewModel.
   at System.ComponentModel.TypeConverter.GetConvertFromException(Object value)
   at System.ComponentModel.TypeConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
   at Microsoft.UI.Xaml.Media.ImageSourceConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value) in /__w/1/s/src/Uno.UI/UI/Xaml/Media/ImageSourceConverter.cs:line 62
   at Uno.UI.DataBinding.BindingPropertyHelper.ConvertUsingTypeDescriptor(Type type, Object value) in /__w/1/s/src/Uno.UI/DataBinding/BindingPropertyHelper.cs:line 1281
   at Uno.UI.DataBinding.BindingPropertyHelper.Convert(Type propertyType, Object value) in /__w/1/s/src/Uno.UI/DataBinding/BindingPropertyHelper.cs:line 1327
   at Microsoft.UI.Xaml.Data.BindingExpression.SetTargetValueSafe(Object v, Boolean useTargetNullValue) in /__w/1/s/src/Uno.UI/DataBinding/BindingExpression.cs:line 747

```

**Comment by HakanL** [comment]
I don't see this error (or anything similar) after upgrading to Uno 6.

**Reply by HakanL** [reply]
I take that back, the logging context had just been changed so I didn't see the errors. The issue is still happening in Uno 6.
